File: android/app/src/main/java/com/oppzippy/openscq30/lib/wrapper/HearId.kt
package com.oppzippy.openscq30.lib.wrapper

import com.oppzippy.openscq30.lib.protobuf.HearId.HearIdCase
import com.oppzippy.openscq30.lib.protobuf.basicHearId
import com.oppzippy.openscq30.lib.protobuf.customHearId
import com.oppzippy.openscq30.lib.protobuf.hearId

sealed class HearId {
    class Basic(val basicHearId: BasicHearId) : HearId()
    class Custom(val customHearId: CustomHearId) : HearId()

    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.HearId {
        return when (this) {
            is Basic -> hearId { basic = basicHearId.toProtobuf() }
            is Custom -> hearId { custom = customHearId.toProtobuf() }
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.HearId.toKotlin(): HearId {
    return when (hearIdCase) {
        HearIdCase.BASIC -> HearId.Basic(basic.toKotlin())
        HearIdCase.CUSTOM -> HearId.Custom(custom.toKotlin())
        HearIdCase.HEARID_NOT_SET -> TODO()
        null -> TODO()
    }
}

data class BasicHearId(
    val isEnabled: Boolean,
    val volumeAdjustments: StereoVolumeAdjustments,
    val time: Int,
) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.BasicHearId {
        return basicHearId {
            isEnabled = this@BasicHearId.isEnabled
            volumeAdjustments = this@BasicHearId.volumeAdjustments.toProtobuf()
            time = this@BasicHearId.time
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.BasicHearId.toKotlin(): BasicHearId {
    return BasicHearId(
        isEnabled = isEnabled,
        volumeAdjustments = volumeAdjustments.toKotlin(),
        time = time,
    )
}

data class CustomHearId(
    val isEnabled: Boolean,
    val volumeAdjustments: StereoVolumeAdjustments,
    val time: Int,
    val hearIdType: UByte,
    val hearIdMusicType: UByte,
    val customVolumeAdjustments: StereoVolumeAdjustments?,
) {
    fun toProtobuf(): com.oppzippy.openscq30.lib.protobuf.CustomHearId {
        return customHearId {
            isEnabled = this@CustomHearId.isEnabled
            volumeAdjustments = this@CustomHearId.volumeAdjustments.toProtobuf()
            time = this@CustomHearId.time
            hearIdType = this@CustomHearId.hearIdType.toInt()
            hearIdMusicType = this@CustomHearId.hearIdMusicType.toInt()
            this@CustomHearId.customVolumeAdjustments?.let {
                customVolumeAdjustments = it.toProtobuf()
            }
        }
    }
}

fun com.oppzippy.openscq30.lib.protobuf.CustomHearId.toKotlin(): CustomHearId {
    return CustomHearId(
        isEnabled = isEnabled,
        volumeAdjustments = volumeAdjustments.toKotlin(),
        time = time,
        hearIdType = hearIdType.toUByte(),
        hearIdMusicType = hearIdMusicType.toUByte(),
        customVolumeAdjustments = if (hasCustomVolumeAdjustments()) customVolumeAdjustments.toKotlin() else null,
    )
}

File: lib/src/devices/standard/implementation/hear_id.rs
use crate::{
    devices::standard::{
        packets::outbound::SetEqualizerAndCustomHearIdPacket,
        state::DeviceState,
        structures::{BasicHearId, CustomHearId, HearId, HearIdMusicType, HearIdType},
    },
    soundcore_device::device::soundcore_command::CommandResponse,
};

pub fn set_hear_id(state: DeviceState, hear_id: HearId) -> crate::Result<CommandResponse> {
    match hear_id {
        HearId::Basic(basic_hear_id) => set_basic_hear_id(state, basic_hear_id),
        HearId::Custom(custom_hear_id) => set_custom_hear_id(state, custom_hear_id),
    }
}

fn set_basic_hear_id(
    state: DeviceState,
    basic_hear_id: BasicHearId,
) -> crate::Result<CommandResponse> {
    let custom_hear_id = CustomHearId {
        is_enabled: basic_hear_id.is_enabled,
        volume_adjustments: basic_hear_id.volume_adjustments.to_owned(),
        // TODO Should this be the current time? If so, what kind of timestamp?
        time: basic_hear_id.time,
        hear_id_type: HearIdType::default(),
        hear_id_music_type: HearIdMusicType::default(),
        custom_volume_adjustments: None,
    };
    let gender = state
        .gender
        .ok_or(crate::Error::MissingData { name: "gender" })?;
    let age_range = state
        .age_range
        .ok_or(crate::Error::MissingData { name: "age range" })?;

    let packet = SetEqualizerAndCustomHearIdPacket {
        equalizer_configuration: &state.equalizer_configuration,
        gender,
        age_range,
        custom_hear_id: &custom_hear_id,
    };

    Ok(CommandResponse {
        packets: vec![packet.into()],
        new_state: DeviceState {
            hear_id: Some(HearId::Basic(basic_hear_id)),
            ..state
        },
    })
}

fn set_custom_hear_id(state: DeviceState, hear_id: CustomHearId) -> crate::Result<CommandResponse> {
    let gender = state
        .gender
        .ok_or(crate::Error::MissingData { name: "gender" })?;
    let age_range = state
        .age_range
        .ok_or(crate::Error::MissingData { name: "age range" })?;
    let packet = SetEqualizerAndCustomHearIdPacket {
        equalizer_configuration: &state.equalizer_configuration,
        gender,
        age_range,
        custom_hear_id: &hear_id,
    };

    Ok(CommandResponse {
        packets: vec![packet.into()],
        new_state: DeviceState {
            hear_id: Some(HearId::Custom(hear_id)),
            ..state
        },
    })
}
